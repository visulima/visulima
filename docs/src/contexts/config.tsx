import { useRouter } from "next/router";
import { ThemeProvider } from "next-themes";
import type { PageOpts } from "nextra";
import { GitHubIcon } from "nextra/icons";
import type { ReactElement, ReactNode } from "react";
import React, {
 createContext, isValidElement, useContext, useMemo, useState,
} from "react";

import Anchor from "../components/anchor";
import FlexSearch from "../components/flexsearch";
import MatchSorterSearch from "../components/match-sorter-search";
import Navbar from "../components/navbar";
import TocPageContent from "../components/toc/toc-page-content";
import TocSidebar from "../components/toc/toc-sidebar";
import { DEFAULT_LOCALE, LEGACY_CONFIG_OPTIONS } from "../constants";
import type { Context, DocumentationThemeConfig } from "../types";
import { getGitEditUrl } from "../utils";
import { MenuProvider } from "./menu";

type Config = DocumentationThemeConfig & Pick<PageOpts, "unstable_flexsearch" | "newNextLinkBehavior" | "title" | "frontMatter">;

const DEFAULT_THEME: DocumentationThemeConfig = {
    banner: {
        dismissible: true,
        key: "nextra-banner",
    },
    darkMode: true,
    direction: "ltr",
    docsRepositoryBase: "https://github.com/shuding/nextra",
    editLink: {
        component({ className, filePath, children }) {
            const editUrl = getGitEditUrl(filePath);
            if (!editUrl) {
                return null;
            }
            return (
                <Anchor className={className} href={editUrl}>
                    {children}
                </Anchor>
            );
        },
        text: "Edit this page",
    },
    feedback: {},
    navbar: {
        component: Navbar,
    },
    footer: {
        copyright: `MIT ${new Date().getFullYear()} © Nextra.`,
    },
    getNextSeoProps: () => {
        return { titleTemplate: "%s – Nextra" };
    },
    gitTimestamp({ timestamp }) {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const { locale = DEFAULT_LOCALE } = useRouter();

        return (
            <>
                Last updated on{" "}
                {timestamp.toLocaleDateString(locale, {
                    day: "numeric",
                    month: "long",
                    year: "numeric",
                })}
            </>
        );
    },
    head: (
        <>
            <meta name="msapplication-TileColor" content="#fff" />
            <meta httpEquiv="Content-Language" content="en" />
            <meta name="description" content="Nextra: the next docs builder" />
            <meta name="twitter:card" content="summary_large_image" />
            <meta name="twitter:site" content="@shuding_" />
            <meta property="og:title" content="Nextra: the next docs builder" />
            <meta property="og:description" content="Nextra: the next docs builder" />
            <meta name="apple-mobile-web-app-title" content="Nextra" />
        </>
    ),
    i18n: [],
    logo: (
        <>
            <span className="font-extrabold">Nextra</span>
            <span className="ml-2 hidden font-normal text-gray-600 md:inline">The Next Docs Builder</span>
        </>
    ),
    logoLink: true,
    navigation: {
        next: true,
        prev: true,
    },
    nextThemes: {
        defaultTheme: "system",
        storageKey: "theme",
    },
    notFound: {
        content: "Submit an issue about broken link",
        labels: "bug",
    },
    primaryHue: {
        dark: 204,
        light: 212,
    },
    project: {
        icon: (
            <>
                <GitHubIcon />
                <span className="sr-only">GitHub</span>
            </>
        ),
    },
    search: {
        component({ className, directories }) {
            // eslint-disable-next-line react-hooks/rules-of-hooks,@typescript-eslint/no-use-before-define
            const config = useConfig();

            return config.unstable_flexsearch ? <FlexSearch className={className} /> : <MatchSorterSearch className={className} directories={directories} />;
        },
        emptyResult: <span className="block select-none p-8 text-center text-sm text-gray-400">No results found.</span>,
        loading() {
            // eslint-disable-next-line react-hooks/rules-of-hooks
            const { locale } = useRouter();

            if (locale === "zh-CN") {
                return "正在加载…";
            }

            if (locale === "ru") {
                return "Загрузка…";
            }

            if (locale === "fr") {
                return "Сhargement…";
            }

            return "Loading…";
        },
        placeholder() {
            // eslint-disable-next-line react-hooks/rules-of-hooks
            const { locale } = useRouter();

            if (locale === "zh-CN") {
                return "搜索文档…";
            }

            if (locale === "ru") {
                return "Поиск документации…";
            }

            if (locale === "fr") {
                return "Rechercher de la documentation…";
            }

            return "Search documentation…";
        },
    },
    serverSideError: {
        content: "Submit an issue about error in url →",
        labels: "bug",
    },
    sidebar: {
        defaultMenuCollapseLevel: 2,
        // eslint-disable-next-line react/jsx-no-useless-fragment
        titleComponent: ({ title }) => <>{title}</>,
    },
    tocSidebar: {
        component: TocSidebar,
        float: true,
        title: "On This Page",
    },
    tocContent: {
        component: TocPageContent,
        float: true,
        title: "On This Page",
    },
};

const DEEP_OBJECT_KEYS = Object.entries(DEFAULT_THEME)
    .map(([key, value]) => {
        const isObject = value && typeof value === "object" && !Array.isArray(value) && !isValidElement(value);

        if (isObject) {
            return key;
        }

        return null;
    })
    .filter(Boolean) as (keyof DocumentationThemeConfig)[];

const theme = {
    title: "",
    frontMatter: {},
    ...DEFAULT_THEME,
};

const ConfigContext = createContext<Config>(theme);

export const useConfig = () => useContext(ConfigContext);

// eslint-disable-next-line radar/cognitive-complexity
export const ConfigProvider = ({ children, value }: { children: ReactNode; value: Context }): ReactElement => {
    const [menu, setMenu] = useState(false);

    const { themeConfig, pageOpts } = value;
    const extendedConfig: Config = useMemo(() => {
        return {
            ...DEFAULT_THEME,
            ...themeConfig,
            unstable_flexsearch: pageOpts.unstable_flexsearch,
            newNextLinkBehavior: pageOpts.newNextLinkBehavior,
            title: pageOpts.title,
            frontMatter: pageOpts.frontMatter,
            ...Object.fromEntries(
                DEEP_OBJECT_KEYS.map((key) => (typeof themeConfig[key] === "object"
                        ? [
                              key,
                              // @ts-expect-error -- key has always object value
                              { ...DEFAULT_THEME[key], ...themeConfig[key] },
                          ]
                        : [])),
            ),
        };
    }, [themeConfig, pageOpts.unstable_flexsearch, pageOpts.newNextLinkBehavior, pageOpts.title, pageOpts.frontMatter]);

    const { nextThemes } = extendedConfig;

    if (process.env.NODE_ENV === "development") {
        const notice = "[nextra-theme-docs] ⚠️  You are using a legacy theme config";

        // TODO: find a good way to refactor this
        // eslint-disable-next-line no-restricted-syntax
        for (const [legacyOption, newPath] of Object.entries(LEGACY_CONFIG_OPTIONS)) {
            if (legacyOption in themeConfig) {
                const [object, key] = newPath.split(".");
                const renameTo = key ? `${object}: { ${key}: ... }` : object;

                // eslint-disable-next-line no-console
                console.warn(`${notice} \`${legacyOption}\`. Rename it to \`${renameTo}\` for future compatibility.`);
            }
        }

        (["search", "footer"] as const).forEach((key) => {
            if (key in themeConfig) {
                const option = themeConfig[key];

                if (typeof option === "boolean" || option == null) {
                    // eslint-disable-next-line no-console
                    console.warn(`${notice} \`${key}\`.`, option ? "Remove it" : `Rename it to \`${key}: { component: null }\` for future compatibility.`);
                }
            }
        });

        if (typeof themeConfig.banner === "string") {
            // eslint-disable-next-line no-console
            console.warn(notice, "`banner`. Rename it to `banner: { content: ... }` for future compatibility.");
        }
    }

    return (
        <ThemeProvider
            attribute="class"
            disableTransitionOnChange
            defaultTheme={nextThemes.defaultTheme}
            storageKey={nextThemes.storageKey}
            forcedTheme={nextThemes.forcedTheme}
        >
            <ConfigContext.Provider value={extendedConfig}>
                <MenuProvider value={{ menu, setMenu }}>{children}</MenuProvider>
            </ConfigContext.Provider>
        </ThemeProvider>
    );
};
