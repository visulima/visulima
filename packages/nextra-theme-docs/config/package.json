{
    "name": "@visulima/nextra-theme-docs-config",
    "version": "0.0.0",
    "description": "A Visulima Nextra theme for documentation sites.",
    "license": "MIT",
    "exports": {
        ".": {
            "types": "./../dist/config.d.ts",
            "require": "./../dist/config.js",
            "import": "./../dist/config.mjs"
        },
        "./package.json": "./package.json"
    },
    "main": "../dist/config.js",
    "module": "../dist/config.mjs",
    "types": "../dist/config.d.ts"
}
