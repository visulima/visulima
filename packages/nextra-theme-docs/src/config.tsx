export { useConfig } from "./contexts";

export { type PartialDocumentsThemeConfig as DocumentationThemeConfig } from "./theme/theme-schema";
