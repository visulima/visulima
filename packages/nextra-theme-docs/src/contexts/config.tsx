import { ThemeProvider } from "next-themes";
import { metaSchema } from "nextra/normalize-pages";
import type { FrontMatter, PageMapItem, PageOpts } from "nextra/types";
import type { ReactElement, ReactNode } from "react";
import { createContext, useContext, useMemo, useState } from "react";
import type { ZodError } from "zod";
import { z } from "zod";

import { DEEP_OBJECT_KEYS, DEFAULT_THEME } from "../constants/default-theme-config";
import type { DocumentationThemeConfig } from "../theme/theme-schema";
import { themeSchema } from "../theme/theme-schema";
import type { Context } from "../types";
import { MenuProvider } from "./menu";

const ConfigContext = createContext<Config>({
    description: "",
    frontMatter: {},
    title: "",
    ...DEFAULT_THEME,
});

let theme: DocumentationThemeConfig;
let isValidated = false;

const extendedMetaSchema = metaSchema.or(
    // eslint-disable-next-line zod/require-strict
    z
        .object({
            description: z.string().optional(),
            icon: z.string().optional(),
            // eslint-disable-next-line zod/require-strict
            theme: z
                .object({
                    prose: z.boolean().optional(),
                })
                .optional(),
        })
        .optional(),
);

const normalizeZodMessage = (error: unknown): string =>
    (error as ZodError).issues
        .flatMap((issue) => {
            const themePath = issue.path.length > 0 && `Path: "${issue.path.join(".")}"`;
            const unionErrors = "unionErrors" in issue ? issue.unionErrors.map((data) => normalizeZodMessage(data)) : [];
            return [[issue.message, themePath].filter(Boolean).join(". "), ...unionErrors];
        })
        .join("\n");

const validateMeta = (pageMap: PageMapItem[]) => {
    pageMap.forEach((pageMapItem) => {
        if (pageMapItem.kind === "Meta") {
            Object.entries(pageMapItem.data).forEach(([key, value]) => {
                try {
                    extendedMetaSchema.parse(value);
                } catch (error) {
                    // eslint-disable-next-line no-console
                    console.error(`[nextra-theme-docs] Error validating _meta.json file for "${key}" property.\n\n${normalizeZodMessage(error)}`);
                }
            });
        } else if (pageMapItem.kind === "Folder") {
            validateMeta(pageMapItem.children);
        }
    });
};

export const useConfig = <FrontMatterType = FrontMatter,>(): Config<FrontMatterType> =>
    // @ts-expect-error TODO: fix Type 'Config<{ [key: string]: any; }>' is not assignable to type 'Config<FrontMatterType>'.
    useContext<Config<FrontMatterType>>(ConfigContext);

export const ConfigProvider = ({ children, value: { pageOpts, themeConfig } }: { children: ReactNode; value: Context }): ReactElement => {
    const [menu, setMenu] = useState(false);

    // Merge only on first load
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    theme ||= {
        ...DEFAULT_THEME,
        ...Object.fromEntries(
            Object.entries(themeConfig).map(([key, value]) => [
                key,
                value && typeof value === "object" && DEEP_OBJECT_KEYS.includes(key)
                    ? // @ts-expect-error -- key has always object value
                      // eslint-disable-next-line security/detect-object-injection
                      { ...DEFAULT_THEME[key], ...(value as object) }
                    : value,
            ]),
        ),
    };

    if (process.env.NODE_ENV !== "production" && !isValidated) {
        try {
            themeSchema.parse(theme);
        } catch (error) {
            // eslint-disable-next-line no-console
            console.error(`[nextra-theme-docs] Error validating theme config file.\n\n${normalizeZodMessage(error)}`);
        }

        validateMeta(pageOpts.pageMap);

        isValidated = true;
    }

    const extendedConfig: Config = useMemo(() => {
        return {
            ...theme,
            description: pageOpts.description,
            flexsearch: pageOpts.flexsearch,
            frontMatter: pageOpts.frontMatter,
            title: pageOpts.title,
        };
    }, [pageOpts]);

    const { darkMode, nextThemes } = extendedConfig;
    const forcedTheme = darkMode ? nextThemes.forcedTheme : "light";

    const menuValue = useMemo(() => {
        return { menu, setMenu };
    }, [menu, setMenu]);

    return (
        <ThemeProvider
            attribute={nextThemes.attribute ?? "class"}
            defaultTheme={nextThemes.defaultTheme}
            disableTransitionOnChange
            forcedTheme={forcedTheme}
            storageKey={nextThemes.storageKey}
        >
            <ConfigContext.Provider value={extendedConfig}>
                <MenuProvider value={menuValue}>{children}</MenuProvider>
            </ConfigContext.Provider>
        </ThemeProvider>
    );
};

export type Config<FrontMatterType = FrontMatter> = DocumentationThemeConfig &
    Pick<PageOpts<FrontMatterType>, "flexsearch" | "frontMatter" | "title"> & { description?: string };
