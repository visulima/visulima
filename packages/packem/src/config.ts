import type { BuildConfig, BuildPreset } from "./types";

export const defineConfig = (config: BuildConfig | BuildConfig[]): BuildConfig[] => (Array.isArray(config) ? config : [config]).filter(Boolean);

export const definePreset = (preset: BuildPreset): BuildPreset => preset;
