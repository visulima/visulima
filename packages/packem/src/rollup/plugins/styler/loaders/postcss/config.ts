import path from "node:path";

import { cosmiconfig } from "cosmiconfig";
import type { AcceptedPlugin, Parser, PluginCreator, Stringifier, Syntax } from "postcss";

import type { PostCSSConfigLoaderOptions } from "../../types";
import { ensurePCSSOption, ensurePCSSPlugins } from "../../utils/options";

interface Options {
    parser?: Parser;
    stringifier?: Stringifier;
    syntax?: Syntax;
}

interface Config extends Options {
    plugins?: (PluginCreator<unknown> | string)[] | Record<string, Record<string, unknown>>;
}

interface Result {
    options: Options;
    plugins: AcceptedPlugin[];
}

export default async function (id: string, config: PostCSSConfigLoaderOptions | false): Promise<Result> {
    if (!config) {
        return { options: {}, plugins: [] };
    }

    const { base, dir, ext } = path.parse(id);

    type Found = { config: Config | ((context: Record<string, unknown>) => Config); isEmpty?: boolean };
    const searchPath = config.path ? path.resolve(config.path) : dir;
    const found: Found | null = await cosmiconfig("postcss").search(searchPath);

    if (!found || found.isEmpty) {
        return { options: {}, plugins: [] };
    }

    const { parser, plugins, stringifier, syntax } =
        typeof found.config === "function"
            ? found.config({
                  cwd: process.cwd(),
                  env: process.env.NODE_ENV ?? "development",
                  file: { basename: base, dirname: dir, extname: ext },
                  options: config.ctx ?? {},
              })
            : found.config;

    const result: Result = { options: {}, plugins: ensurePCSSPlugins(plugins) };

    if (parser) {
        result.options.parser = ensurePCSSOption(parser, "parser");
    }

    if (syntax) {
        result.options.syntax = ensurePCSSOption(syntax, "syntax");
    }

    if (stringifier) {
        result.options.stringifier = ensurePCSSOption(stringifier, "stringifier");
    }

    return result;
}
