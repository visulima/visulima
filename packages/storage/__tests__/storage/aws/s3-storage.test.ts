import {
    CompleteMultipartUploadCommand,
    CopyObjectCommand,
    CreateMultipartUploadCommand,
    DeleteObjectCommand,
    HeadObjectCommand,
    ListPartsCommand,
    S3Client,
    UploadPartCommand,
} from "@aws-sdk/client-s3";
import { mockClient } from "aws-sdk-client-mock";
import { createRequest } from "node-mocks-http";
import { beforeEach, describe, expect, expectTypeOf, it, vi } from "vitest";

import S3Storage from "../../../src/storage/aws/s3-storage";
import type { AwsError, S3StorageOptions } from "../../../src/storage/aws/types";
import { metafile, storageOptions, testfile } from "../../__helpers__/config";

vi.mock(import("aws-crt"));
vi.mock("@aws-sdk/s3-request-presigner", () => ({
    getSignedUrl: vi.fn().mockResolvedValue("https://api.s3.example.com?signed"),
}));

const s3Mock = mockClient(S3Client);

describe(S3Storage, () => {
    let getSignedUrlMock: any;
    vi.useFakeTimers().setSystemTime(new Date("2022-02-02"));

    const options = { ...(storageOptions as S3StorageOptions), bucket: "bucket", region: "us-east-1" };

    const request = createRequest();

    let storage: S3Storage;

    const metafileResponse = {
        Metadata: {
            metadata: encodeURIComponent(
                JSON.stringify({
                    ...metafile,
                    bytesWritten: 0,
                    createdAt: new Date().toISOString(),
                    status: "created",
                    UploadId: "987654321",
                }),
            ),
        },
    };

    beforeEach(async () => {
        s3Mock.reset();
        storage = new S3Storage(options);

    });

    describe(".create()", () => {
        it("should request api and set status and UploadId", async () => {
            s3Mock.on(HeadObjectCommand).rejects();
            s3Mock.on(CreateMultipartUploadCommand).resolves({ UploadId: "123456789" });

            const s3file = await storage.create(request, metafile);

            expect(s3file).toMatchSnapshot({
                expiredAt: expect.any(Number),
                createdAt: expect.any(String),
            });
        });

        it("should handle existing", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);

            const s3file = await storage.create(request, metafile);

            expect(s3file).toMatchSnapshot();
        });

        it("should send error on invalid s3 response", async () => {
            s3Mock.on(HeadObjectCommand).rejects();
            s3Mock.on(CreateMultipartUploadCommand).resolves({});

            await expect(storage.create(request, metafile)).rejects.toMatchSnapshot();
        });

        it("should handle TTL option", async () => {
            s3Mock.on(HeadObjectCommand).rejects();
            s3Mock.on(CreateMultipartUploadCommand).resolves({ UploadId: "123456789" });

            const s3file = await storage.create(request, { ...metafile, ttl: "30d" });

            expect(s3file.expiredAt).toBeDefined();

            expectTypeOf(s3file.expiredAt).toBeNumber();

            // TTL should be converted to expiredAt timestamp
            const expectedExpiry = Date.now() + 30 * 24 * 60 * 60 * 1000; // 30 days in ms

            expect(s3file.expiredAt).toBeGreaterThan(expectedExpiry - 1000); // Allow 1s tolerance
            expect(s3file.expiredAt).toBeLessThan(expectedExpiry + 1000);
        });
    });

    describe(".update()", () => {
        it("should update changed metadata keys", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);

            // eslint-disable-next-line radar/no-duplicate-string
            const s3file = await storage.update(metafile, { metadata: { name: "newname.mp4" } });

            expect(s3file.metadata.name).toBe("newname.mp4");
            expect(s3file.metadata.mimeType).toBe(metafile.metadata.mimeType);
        });

        it("should reject if not found", async () => {
            expect.assertions(1);

            await expect(storage.update(metafile, { metadata: { name: "newname.mp4" } })).rejects.toHaveProperty("UploadErrorCode", "FileNotFound");
        });

        it("should handle TTL option in update", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);

            const s3file = await storage.update(metafile, { ttl: "1h" });

            expect(s3file.expiredAt).toBeDefined();

            expectTypeOf(s3file.expiredAt).toBeNumber();

            // TTL should be converted to expiredAt timestamp
            const expectedExpiry = Date.now() + 60 * 60 * 1000; // 1 hour in ms

            expect(s3file.expiredAt).toBeGreaterThan(expectedExpiry - 1000); // Allow 1s tolerance
            expect(s3file.expiredAt).toBeLessThan(expectedExpiry + 1000);
        });
    });

    describe(".write()", () => {
        it("should request api and set status and bytesWritten", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);
            s3Mock.on(ListPartsCommand).resolves({ Parts: [] });
            s3Mock.on(UploadPartCommand).resolves({ ETag: "1234" });
            s3Mock.on(CompleteMultipartUploadCommand).resolves({ Location: "/1234" });

            const part = {
                body: testfile.asReadable,
                contentLength: metafile.size,
                id: metafile.id,
                name: metafile.name,
                start: 0,
            };
            const s3file = await storage.write(part);

            expect(s3file).toMatchSnapshot();
        });

        it("should request api and set status and bytesWritten on resume", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);
            s3Mock.on(ListPartsCommand).resolves({ Parts: [] });

            const part = {
                contentLength: 0,
                id: metafile.id,
                name: metafile.name,
            };
            const s3file = await storage.write(part);

            expect(s3file).toMatchSnapshot();
        });
    });

    describe("delete()", () => {
        it("should set status", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);
            s3Mock.on(DeleteObjectCommand).resolves({});

            const deleted = await storage.delete(metafile);

            expect(deleted.status).toBe("deleted");
        });

        it("should ignore if not exist", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);

            const deleted = await storage.delete(metafile);

            expect(deleted.id).toBe(metafile.id);
        });
    });

    describe("copy()", () => {
        it("relative", async () => {
            s3Mock.resetHistory();
            s3Mock.on(CopyObjectCommand).resolves({});

            await storage.copy("name", "new name");

            expect(s3Mock.call(0).args[0].input).toStrictEqual({
                Bucket: "bucket",
                CopySource: "bucket/name",
                Key: "new name",
            });
        });

        it("absolute", async () => {
            s3Mock.resetHistory();
            s3Mock.on(CopyObjectCommand).resolves({});

            await storage.copy("name", "/otherBucket/new name");

            expect(s3Mock.call(0).args[0].input).toStrictEqual({
                Bucket: "otherBucket",
                CopySource: "bucket/name",
                Key: "new name",
            });
        });

        it("with storage class", async () => {
            s3Mock.resetHistory();
            s3Mock.on(CopyObjectCommand).resolves({});

            await storage.copy("name", "new name", { storageClass: "GLACIER" });

            expect(s3Mock.call(0).args[0].input).toStrictEqual({
                Bucket: "bucket",
                CopySource: "bucket/name",
                Key: "new name",
                StorageClass: "GLACIER",
            });
        });
    });

    describe("normalizeError", () => {
        it("s3 error", () => {
            const error = {
                $metadata: { httpStatusCode: 400 },
                Code: "SomeServiceException",
                message: "SomeServiceException",
                name: "SomeError",
            };

            expect(storage.normalizeError(error)).toMatchSnapshot();
        });

        it("not s3 error", () => {
            expect(storage.normalizeError(new Error("unknown") as AwsError)).toMatchSnapshot();
        });
    });
});

describe("s3PresignedStorage", () => {
    let getSignedUrlMock: any;
    vi.useFakeTimers().setSystemTime(new Date("2022-02-02"));

    const options = {
        ...(storageOptions as S3StorageOptions),
        bucket: "bucket",
        clientDirectUpload: true,
        region: "us-east-1",
    };

    const request = createRequest();

    let storage: S3Storage;

    const metafileResponse = {
        Metadata: {
            metadata: encodeURIComponent(
                JSON.stringify({
                    ...metafile,
                    bytesWritten: 0,
                    createdAt: new Date().toISOString(),
                    status: "created",
                    UploadId: "987654321",
                }),
            ),
        },
    };

    beforeEach(async () => {
        s3Mock.reset();
        storage = new S3Storage(options);

    });

    describe(".create()", () => {
        it("should request api and set status and UploadId", async () => {
            s3Mock.on(HeadObjectCommand).rejects();
            s3Mock.on(CreateMultipartUploadCommand).resolves({ UploadId: "123456789" });
            s3Mock.on(ListPartsCommand).resolves({ Parts: [] });

            const s3file = await storage.create(request, metafile);

            expect(s3file.partsUrls?.length).toBe(1);
            expect(s3file.partSize).toBeGreaterThan(0);
        });
    });

    describe("update", () => {
        it("should add partsUrls", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);
            s3Mock.on(ListPartsCommand).resolves({ Parts: [] });
            const s3file = await storage.update(metafile, metafile);

            expect(s3file.partsUrls?.length).toBe(1);
        });

        it("should complete", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);

            const preCompleted = {
                ...metafile,
                Parts: [{ ETag: "123456789", PartNumber: 1, Size: 64 }],
                partSize: 16_777_216,
                partsUrls: ["https://api.s3.example.com?signed"],
                UploadId: "123456789",
            };

            s3Mock.on(CompleteMultipartUploadCommand).resolves({ Location: "/1234" });

            const s3file = await storage.update({ id: metafile.id }, preCompleted);

            expect(s3file.status).toBe("updated");
        });

        it("should complete (empty payload)", async () => {
            s3Mock.on(HeadObjectCommand).resolves(metafileResponse);
            s3Mock.on(ListPartsCommand).resolves({ Parts: [{ ETag: "123456789", PartNumber: 1, Size: 64 }] });
            s3Mock.on(CompleteMultipartUploadCommand).resolves({ Location: "/1234" });

            const s3file = await storage.update({ id: metafile.id }, {});

            expect(s3file.status).toBe("updated");
        });
    });
});
