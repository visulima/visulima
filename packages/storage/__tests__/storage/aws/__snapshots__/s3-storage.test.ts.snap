// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`s3 > .create() > should handle existing 1`] = `
{
  "UploadId": "987654321",
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
}
`;

exports[`s3 > .create() > should request api and set status and UploadId 1`] = `
{
  "ETag": undefined,
  "Parts": undefined,
  "UploadId": "123456789",
  "bytesWritten": 0,
  "content": undefined,
  "contentType": "video/mp4",
  "createdAt": Any<String>,
  "expiredAt": Any<Number>,
  "hash": undefined,
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "modifiedAt": undefined,
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "partSize": undefined,
  "partsUrls": undefined,
  "size": 64,
  "status": "created",
  "uri": undefined,
}
`;

exports[`s3 > .create() > should send error on invalid s3 response 1`] = `[UploadError: s3 create upload error]`;

exports[`s3 > .write() > should request api and set status and bytesWritten 1`] = `
{
  "ETag": undefined,
  "UploadId": "987654321",
  "bytesWritten": 64,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "completed",
  "uri": "/1234",
}
`;

exports[`s3 > .write() > should request api and set status and bytesWritten on resume 1`] = `
{
  "Parts": [],
  "UploadId": "987654321",
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "part",
}
`;

exports[`s3 > normalizeError > not s3 error 1`] = `
{
  "code": "Error",
  "message": "[s3] unknown",
  "name": "Error",
  "statusCode": 500,
}
`;

exports[`s3 > normalizeError > s3 error 1`] = `
{
  "code": "SomeServiceException",
  "message": "SomeServiceException",
  "name": "SomeError",
  "statusCode": 400,
}
`;
